### PR TITLE
 Add 'Save' function to boards #781 

### LINKED ISSUE #781: Add 'Save' function to boards
Currently, the 'Save' function works as 'Save as'. Rename it and add a Save function to update the board under the current name.

### PR DESCRIPTION
**Pull request by HyungJon**
Fixes #781 

"Save" function now refers to saving the state of the last opened board under its current name. The old function has been renamed to "Save as".
Saving when there is no board open results in the same behavior as "Save as".

### CODE REVIEW

**Comment by HyungJon** [comment]
Unexpected Travis failure again. This time I will squash noisy commits when it's fixed.

**Comment by HyungJon** [comment]
All right, I think the last push has been squashed properly this time. At least they show up as squashed in Git log on my local repository.

File: src/main/java/prefs/GlobalConfig.java [hunk 1]
```diff
@@ -28,6 +28,7 @@
     private String lastViewedRepository = "";
     private String lastLoginUsername = "";
     private byte[] lastLoginPassword = new byte[0];
+    private String lastOpenBoard = "";
```
Comment (dariusf) [inline src/main/java/prefs/GlobalConfig.java]: We should use `Optional<String>` here instead. Expresses the semantics of this field more clearly.

Also, the empty string is a valid board name... Probably shouldn't be, but its use here as an empty value could cause problems.

File: src/main/java/ui/MenuControl.java [hunk 2]
```diff
@@ -32,12 +32,14 @@
     private final ScrollPane panelsScrollPane;
     private final UI ui;
     private final Preferences prefs;
+    private String openBoardName = "";
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: Is there a need to keep track of this here as well, i.e. could this and the state in prefs ever go out of sync?

File: src/main/java/ui/MenuControl.java [hunk 3]
```diff
@@ -129,12 +131,31 @@ public void changed(ObservableValue<? extends Number> arg0, Number arg1, Number
...
     }
+    
+    private void onBoardSave() {
+        logger.info("Menu: Boards > Save");
+        
+        List<PanelInfo> panels = getCurrentPanels();
+        
+        if (openBoardName.equals("")) {
+            onBoardSaveAs();
+        }
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: Should there be a `return` here? Seems like after we go into `onBoardSaveAs` we'd either have saved the board or not, and we shouldn't have to come back here.

File: src/main/java/prefs/GlobalConfig.java (hunk 1 above)
Comment (HyungJon) [inline src/main/java/prefs/GlobalConfig.java]: Hmm true, I should add some name validation here.

File: src/main/java/ui/MenuControl.java (hunk 3 above)
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: Also, if there should be a return, you could swap the conditional and the `getCurrentPanels()` line, to make it clear that the absence of a board name immediately delegates to `onBoardSaveAs` without needing to do anything else.

File: src/main/java/prefs/GlobalConfig.java [hunk 4]
```diff
@@ -82,6 +84,14 @@ public void addBoard(String name, List<PanelInfo> panels) {
     public void removeBoard(String name) {
         savedBoards.remove(name);
     }
+    
+    public void setLastOpenBoard(String board) {
+        lastOpenBoard = Optional.of(board);
+    }
+    
+    public String getLastOpenBoard() {
+        return lastOpenBoard.get();
```
Comment (dariusf) [inline src/main/java/prefs/GlobalConfig.java]: The point of `Optional` is to express statically that a value may be absent, and force consumers of the value to check for this (unlike `null`, which doesn't make you check anything).

Calling `get` without checking `isPresent` will fail at runtime (similar to dereferencing a null pointer). You should instead propagate the `Optional` out to callers of this method.

File: src/main/java/prefs/GlobalConfig.java [hunk 5]
```diff
@@ -82,6 +84,14 @@ public void addBoard(String name, List<PanelInfo> panels) {
     public void removeBoard(String name) {
         savedBoards.remove(name);
     }
+    
+    public void setLastOpenBoard(String board) {
```
Comment (dariusf) [inline src/main/java/prefs/GlobalConfig.java]: Similarly, there needs to be a way for callers to indicate that there is no last-opened board. You could either have it take an `Optional` or provide `clearLastOpenBoard`.

File: src/main/java/ui/MenuControl.java [hunk 6]
```diff
@@ -129,12 +129,31 @@ public void changed(ObservableValue<? extends Number> arg0, Number arg1, Number
         cols.getItems().addAll(createRight, createLeft, closePanel);
         return cols;
     }
+    
+    private void onBoardSave() {
+        logger.info("Menu: Boards > Save");
+        
+        if (prefs.getLastOpenBoard().equals("")) {
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: Now that callers get an `Optional`, they can check if it `isPresent`.

File: src/main/java/ui/MenuControl.java [hunk 7]
```diff
@@ -182,6 +208,9 @@ private void onBoardDelete(String boardName) {
 
         if (response.isPresent() && response.get().getButtonData() == ButtonData.OK_DONE) {
             prefs.removeBoard(boardName);
+            if (prefs.getLastOpenBoard().equals(boardName)) {
+                prefs.setLastOpenBoard("");
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: This is where that `clear` method would be useful.

We also have to handle the possibility of the last-opened board not being present.

File: src/main/java/ui/MenuControl.java [hunk 8]
```diff
@@ -151,9 +170,15 @@ private void onBoardSave() {
         Optional<String> response = dlg.showAndWait();
 
         if (response.isPresent()) {
-            prefs.addBoard(response.get(), panels);
+            String boardName = response.get();
+            if (boardName.equals("")) {
+                logger.info("Did not save new board");
+                return;
+            }
```
Comment (dariusf) [inline src/main/java/ui/MenuControl.java]: You might want to factor the validation out into another method, and log each case, for example if it's empty, if it consists only of whitespace, etc.

**Comment by HyungJon** [comment]
Maybe I should look into the ```NoNodesException``` again.

File: src/main/java/prefs/GlobalConfig.java [hunk 9]
```diff
@@ -28,6 +29,7 @@
     private String lastViewedRepository = "";
     private String lastLoginUsername = "";
     private byte[] lastLoginPassword = new byte[0];
+    private Optional<String> lastOpenBoard;
```
Comment (dariusf) [inline src/main/java/prefs/GlobalConfig.java]: Think you might have forgotten to initialise this (to `empty`). Got an NPE when I checked this out and tried to save. It wasn't caught in the test?

**Comment by HyungJon** [comment]
Some ridiculous error causing Travis failure, somehow trying to fine "'Node should have label Save as'" instead of just "Save as". I'll look into squashing these commits as well after I resolve them.

**Comment by dariusf** [comment]
No worries, just ping or change the label when you are done.

**Comment by HyungJon** [comment]
I just changed the tests to select all menu items using arrow keys. At least temporarily.

**Comment by dariusf** [comment]
Hmm, using arrow keys might be fragile if we add more menu items. @lejolly is there another way to avoid the problem?

Otherwise this looks good.

**Comment by damithc** [comment]
@lejolly see darius' comment above.

**Comment by lejolly** [comment]
Whoops, missed this, I don't think the ci works well with clicking on menus. The alternative would be to use the accelerator (keyboard shortcut).

**Comment by lejolly** [comment]
I guess the upside of using arrow keys is that we know if the order of items is kept.

**Comment by damithc** [comment]
That should be a separate issue/PR. Saving borads is not frequent, so we may not need a shortcut for that. But I would like to have a short cut for changing boards, similar to the one we have for changing repos.

**Comment by HyungJon** [comment]
All right. Will Ctrl+B do, like Ctrl+R for changing repos?

**Comment by damithc** [comment]
Create a new issue. We'll discuss it over there.

**Comment by dariusf** [comment]
In future you could squash a bit more. Build failures, 'fixed bug'-type commits, review changes (i.e. things which should have been done to begin with) can be part of whichever commit they are most related to.

Test changes can be a single commit, unless there's some change in there that warrants more granular documentation.

**Comment by HyungJon** [comment]
Noted, I'll make sure to squash commits as required before the pushing itself.

**Pull request merged**